Standard/src/Synthesis/UnitaryDecomposition.cs

```diff
@@ -0,0 +1,172 @@
+                        }
+                        var u2x2 = new TwoLevelUnitary(mx, j - 1, j);
+                        u2x2.MultiplyRight(A);
+                        u2x2.ConjugateTranspose();
+                        if (!u2x2.IsIdentity())
+                        {
+                            result.Add(u2x2);
+                        }
+                    }
+
+                    // At this point all elements in row i and in column i are zeros, with exception
+                    // of diagonal element A[i, i], which is equal to 1.
+                    Debug.Assert((A[i, i] - 1.0).Magnitude < tol);
+                }
+
+                var lastMatrix = new TwoLevelUnitary(new Complex[,] {
+                    { A[n - 2, n - 2], A[n - 2, n - 1] },
+                    { A[n - 1, n - 2], A[n - 1, n - 1] } }, n - 2, n - 1);
+                if (!lastMatrix.IsIdentity())
+                {
+                    result.Add(lastMatrix);
+                }
+                return result;
+            }
+
+            // Returns permutation of numbers from 0 to n-1 such that any two consequent numbers in
+            // it differ in exactly one bit.
+            private static int[] GrayCode(int n)
+            {
+                var result = new int[n];
+                for (int i = 0; i < n; i++)
+                {
+                    result[i] = i ^ (i / 2);
+                }
+                return result;
+            }
+
+            // Applies permutation to columns and rows of matrix.
+            private static Complex[,] PermuteMatrix(Complex[,] matrix, int[] perm)
+            {
+                int n = perm.Length;
+                Debug.Assert(matrix.GetLength(0) == n);
+                Debug.Assert(matrix.GetLength(1) == n);
+                var result = new Complex[n, n];
+                for (int i = 0; i < n; i++)
+                {
+                    for (int j = 0; j < n; j++)
+                    {
+                        result[i, j] = matrix[perm[i], perm[j]];
+                    }
+                }
+                return result;
+            }
+
+            // Returns list of two-level unitary matrices, which multiply to A.
+            // Every matrix has indices differing in exactly 1 bit.
+            private static List<TwoLevelUnitary> twoLevelDecomposeGray(Complex[,] A)
+            {
+                int n = A.GetLength(0);
+                Debug.Assert(A.GetLength(1) == n, "Matrix is not square.");
+                Debug.Assert(MatrixUtils.IsMatrixUnitary(A), "Matrix is not unitary.");
+
+                int[] perm = GrayCode(n);
+                A = PermuteMatrix(A, perm);
+                List<TwoLevelUnitary> result = TwoLevelDecompose(A);
+                foreach (TwoLevelUnitary matrix in result)
+                {
+                    matrix.ApplyPermutation(perm);
+                    matrix.OrderIndices();
+                }
+                return result;
+            }
```

Comment:
You may not need to collect in a `List` here, either. Using the `yield return` statement (very similar to Python's `yield`), you can write this out directly as an enumerable over `TwoLevelUnitary`:

```suggestion
            private static IEnumerable<TwoLevelUnitary> twoLevelDecomposeGray(Complex[,] A)
            {
                int n = A.GetLength(0);
                Debug.Assert(A.GetLength(1) == n, "Matrix is not square.");
                Debug.Assert(MatrixUtils.IsMatrixUnitary(A), "Matrix is not unitary.");

                int[] perm = GrayCode(n);
                A = PermuteMatrix(A, perm);
                foreach (TwoLevelUnitary matrix in  TwoLevelDecompose(A))
                {
                    matrix.ApplyPermutation(perm);
                    matrix.OrderIndices();
                    yield return matrix;
                }
            }
```

Using the same technique in `TwoLevelDecompose` should allow you to avoid allocating a list for any intermediate steps, only allocating in the constructor call at line 168.

Reply:
Done.